source/Halibut/Util/AsyncEx/TaskExtensions.cs

```diff
@@ -0,0 +1,73 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Halibut.Util.AsyncEx
+{
+    internal static class TaskExtensions
+    {
+        /// <summary>
+        /// Waits for a) the task to complete, b) the timeout to pass or c) the cancellationToken to be triggered.
+        /// Wraps it all up so that we dont get UnobservedTaskExceptions 
+        /// </summary>
+        /// <param name="task">The task to wait on</param>
+        /// <param name="timeout">The amount of time to wait until we give up</param>
+        /// <param name="cancellationToken">Supports task cancellation</param>
+        /// <returns>The task if successful, otherwise a TimeoutException or OperationCanceledException</returns>
+        /// <exception cref="TimeoutException">If the timeout gets reached before the task completes</exception>
+        /// <exception cref="OperationCanceledException">The task was cancelled via the cancellation token</exception>
+        public static async Task TimeoutAfter(this Task task, TimeSpan timeout, CancellationToken cancellationToken)
+        {
+            var timeOutTask = Task.Delay(timeout);
+            var cancellationTask = cancellationToken.AsTask();
+            var timeoutCancellation = new CancellationTokenSource();
```

Comment:
This should be disposed: https://docs.microsoft.com/en-us/dotnet/api/system.threading.cancellationtokensource.dispose?view=netframework-4.8